### PR TITLE
[SPARK-26439][CORE][WIP] Introduce WorkerOffer reservation mechanism for Barrier TaskSet

### PR DESCRIPTION
**Pull request by Ngone51**
## What changes were proposed in this pull request?

Currently, Barrier TaskSet has a hard requirement that tasks can only be launched
in a single resourceOffers round with enough slots(or sufficient resources), but
can not be guaranteed even if with enough slots due to task locality delay scheduling.
So, it is very likely that Barrier TaskSet gets a chunk of sufficient resources after
all the trouble, but let it go easily just because one of pending tasks can not be
scheduled. Futhermore, it causes severe resource competition between TaskSets and jobs
and introduce unclear semantic for DynamicAllocation.

This pr trys to introduce WorkerOffer reservation mechanism for Barrier TaskSet, which
allows Barrier TaskSet to reserve WorkerOffer in each resourceOffers round, and launch
tasks at the same time once it accumulate the sufficient resource. In this way, we 
relax the requirement of resources for the Barrier TaskSet. 

Besides, we have two features along with  WorkerOffer reservation mechanism:

* To avoid the deadlock which may be introuduced by serveral Barrier TaskSets holding the reserved WorkerOffers for a long time, we'll ask Barrier TaskSets to force releasing part of reserved WorkerOffers
on demand. So, it is highly possible that each Barrier TaskSet would be launched in the end.

* Barrier TaskSet could replace old high level locality reserved WorkerOffer with new low level locality WorkerOffer during the time it wating for sufficient resources, to perform better locality at the end.

To integrate with DynamicAllocation:

The possible effective way I can imagine is that adding new event, e.g. 
ExecutorReservedEvent, ExecutorReleasedEvent, which behaved like busy executor with
running tasks or idle executor without running tasks. Thus, ExecutorAllocationManager 
would not let the executor go if it reminds of there're some reserved resource on that
executor.

## How was this patch tested?

existed and added some, needs to add more.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -715,6 +715,24 @@ package object config {
...
       .createWithDefault(40)
 
+  private[spark] val BARRIER_NO_SUFFICIENT_RESOURCE_TIMEOUT =
+    ConfigBuilder("spark.scheduler.barrier.noSufficientResource.timeout")
+      .doc("Time in minutes to wait before a barrier TaskSet get sufficient resource " +
+        "to launch tasks. Abort the barrier TaskSet once it expires to avoid job " +
+        "hanging indefinitely.")
+      .timeConf(TimeUnit.MINUTES)
+      .checkValue(v => v > 0, "Time value should be a positive value.")
+      .createWithDefault(5)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: Not sure what's the appreciate value here, just an arbitrary value right now.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 2]
```diff
@@ -715,6 +715,24 @@ package object config {
...
+
+  private[spark] val BARRIER_MAX_CONSECUTIVE_NO_BARRIER_TASKSET_LAUNCH_TIMES =
+    ConfigBuilder("spark.scheduler.barrier.maxConsecutiveNoBarrierTaskSetLaunchTimes")
+      .doc("Number of max consecutive times of no any barrier taskSet launched in each " +
+        "resourceOffers round. TaskScheduler will ask barrier taskSets to release reserved" +
+        "WorkOffers if it reach this point.")
+      .intConf
+      .checkValue(v => v > 0, "The maxConsecutiveNoTaskSetLaunchTimes should be a positive value.")
+      .createWithDefault(5)
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala]: ditto

**Comment by Ngone51** [comment]
ping @squito @cloud-fan @jiangxb1987 

Would you mind taking a look at this pr ? Thanks.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 3]
```diff
@@ -303,45 +316,136 @@ private[spark] class TaskSchedulerImpl(
...
+                    // given the WorkerOffer a second chance to offer the resource for the taskSet,
+                    // which reclaims the resource just now
+                    addedReviveOffersPerRound += shuffledOffers(offerIndex)
+                  }
+                }
+                // don't do another resourceOffer round if we've already archived the goal, even if
+                // we have new added reviveOffers in this round, which may provide better locality
+                // preference, but that's not guaranteed.
+                if (taskSet.getReadyTaskToReservedWorkerOffer.size == taskSet.numTasks) {
+                  return launchedTask
```
Comment (Ngone51) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Maybe, we should remove this to perform a better locality scheduler.

**Review by felixcheung — COMMENTED** [review]
this can be interesting.. but isn't the allocation is going to fail if the schedulerbackend does't have enough resource for executor?

**Comment by Ngone51** [comment]
Hi, @felixcheung , thanks for your comment.  Hmmm....I'm not pretty sure what you mean here, so,  just given my understanding here. 

As far as I know, schedulerbackend is not responsible for resource management. It just request executors to cluster manager, but do not know clearly about the available resources.

And for Barrier TaskSet, if app's maximum config resources is less than Barrier TaskSet's demand, we'll fail the related stage, or even job at the end. You can see `DAGScheduler.checkBarrierStageWithNumSlots` for details.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 4]
```diff
@@ -382,100 +475,97 @@ private[spark] class TaskSchedulerImpl(
...
     // NOTE: the preferredLocality order: PROCESS_LOCAL, NODE_LOCAL, NO_PREF, RACK_LOCAL, ANY
     for (taskSet <- sortedTaskSets) {
-      // Skip the barrier taskSet if the available slots are less than the number of pending tasks.
-      if (taskSet.isBarrier && availableSlots < taskSet.numTasks) {
-        // Skip the launch process.
-        // TODO SPARK-24819 If the job requires more slots than available (both busy and free
-        // slots), fail the job on submit.
-        logInfo(s"Skip current round of resource offers for barrier stage ${taskSet.stageId} " +
-          s"because the barrier taskSet requires ${taskSet.numTasks} slots, while the total " +
-          s"number of available slots is $availableSlots.")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: This msg seems to be lost.  I think this is related to felix's point -- though there wasn't a hard failure here before, at least there was some info to the user before, but now it seems that the user won't get any info.

**Review by squito — COMMENTED** [review]
I've only looked at this briefly so far. I appreciate work being done to improve barrier scheduling, but I'm a hesitant on such a big change to the scheduler.  IIUC, by itself the improvement here is just for having multiple concurrent barrier stages, still with static allocation?  I can see how this could eventually help with dynamic allocation, but also not confident about how complex that is going to end up being.

Have you looked at gang scheduling in yarn?  I don't know how that works, but we should at least compare with it.

I think for dynamic allocation & barrier mode to really work well together, we'd need to tie into the gang-scheduling of the resource managers, so the long-term plan needs to consider that.

I think at this stage, we might benefit more from a high-level design discussion than from code?

**Comment by Ngone51** [comment]
> such a big change to the scheduler

It may be a big change for Barrier TaskSet, but remains no difference for non-barrier TaskSet yet.

 > by itself the improvement here is just for having multiple concurrent barrier stages, still with static allocation?

Yes, and I think the improvement here could also handle a mix sitution of non-barrier & barrier TaskSets.
Dynamic allocation would be the next step if this change could be accepted.

> Have you looked at gang scheduling in yarn?

Sorry, no. But I know YARN has reservation mechanism for resource, too. I'll have a look for YARN gang scheduling. 

> I think at this stage, we might benefit more from a high-level design discussion than from code?

Agreed. And it would be better if we could get more people in. Let me try @jiangxb1987 @jerryshao

**Comment by Ngone51** [comment]
Hi @squito , I just did a shallow research into YARN gang scheduling.
 As far as I relazied, YARN do not have a public API directly for gang scheduling, but using its Reservation System(which [have gang semantics](http://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/ReservationSystem.html)) to work around gang scheduing (but is not designed for it) at **AM** level. And there's key difference that *[The goal of barrier scheduler is to offer gang semantics in the task scheduling level, whereas the gang semantics in the YARN level is more regarding to resource level.](https://issues.apache.org/jira/browse/SPARK-24723?focusedCommentId=16533205&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16533205)*

**Comment by squito** [comment]
Thanks for the links @Ngone51 .  I've briefly looked through them, but I think I would need more time to really think through all of this.  At least in the near future, I don't think I'm going to have a ton of time to look at this (sorry I know that isn't helpful, but figured I should let you know).

**Comment by Ngone51** [comment]
It's okay and your feedback would be welcome at any time. Any way, thanks a lot, @squito .

**Comment by felixcheung** [comment]
thanks @Ngone51 - I think you did bring up some very good point here in your description, and would be precisely what I think we would need to get barrier scheduling actually useful for its intention use cases. would you mind thinking about a way this could be done with smaller changes? also welcome to bring the design discussion to dev@spark.apache.org

**Comment by Ngone51** [comment]
@felixcheung Thanks for your suggestion, I'll try it.

**Pull request closed without merging**